clang/test/CIR/CodeGen/float16-arithmetic.c

```diff
@@ -0,0 +1,99 @@
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fclangir -emit-cir %s -o %t.cir
+// RUN: FileCheck --input-file=%t.cir %s
+
+_Float16 add1(_Float16 a, _Float16 b) {
+  return a + b;
+}
```

Comment:
16-bit fp requires different LLVM lowering for different arches, example: ARM64 vs x86_64. I'm a bit skeptical this PR will just automatically work properly (I don't see any arch specific checks, and perhaps there's some ABI work missing?), quick examples:
x86_64: https://godbolt.org/z/vr4hG6v3x
ARM64: https://godbolt.org/z/Mn8znbbhx

Handwaving to proper support right now might give us the wrong impression that this actually work, where in truth it doesn't and will require folks to spend hours investigating why it does not. What I'd like to see here:
- LLVM checking for one at least one Arch.
- Should some of this arch specific things be handled in CIR already? I'd expect this to already be the case for other floating point types, which makes me worried that current fp support took the shortest path and might not actually be right when we lower down to LLVM (e.g. fp80) - I hope I'm wrong cause that's gonna be hard to pinpoint when we get correctness issues while running code.
- Somehow make sure that LowerToLLVM.cpp (or CIRGen if details get exposed to that level) gives us error/asserts for arches it doesn't know how to do the right job.

Reply:
I checked the upstream clang CodeGen code. The relevant code is:

https://github.com/llvm/clangir/blob/981c976f692c1618714348967ddaca55e99ce76e/clang/lib/CodeGen/CGExprScalar.cpp#L841-L847

The basic idea is, before emitting a unary or binary scalar expression of floating point type, CodeGen checks whether the operands should be promoted by calling `Ty.UseExcessPrecision`. If this call returns `true`, the operands are promoted to `float` before evaluating the expression. The call is target-specific, on x86 it returns `true` and on `aarch64` it returns `false`. So the promotion of the operands need to be handled similarly at the CIR side if we're going to align to the upstream clang CodeGen. Here is a possible approach:

- Invent a new attribute that holds global floating point related configurations (such as whether half-precision types should be promoted), and attach this attribute to the `module` operation, just like `#cir.lang` and `#cir.sob`.
- During LLVM lowering prepare, teach the transformation pass to promote the operands of `cir.unary` and `cir.binop` accordingly from the information given by the new module attribute.

This may be a non-trivial change. Any thoughts on this?

Reply:
Thanks for pointing out the difference in terms of code. Seems like some of my previous PR reviews slip through the cracks on letting these operations to be emitted without proper calls to `getPromotionType(...);`.

I'm a bit confused on why do you think this needs to be done post CIRGen - we need to do these promotions/extensions during CIRGen (by calling `getPromotionType`, etc), I don't see any current advantage of postponing these specific ABI lowerings.

I'm coming from the correctness angle here, if things like this sneak in without proper correctness, it will be a lot of work to investigate these issues during runtime later on. I appreciate your work on adding these fp variants but we first need to make sure that we have a sound path (or at least enough asserts for things we don't yet handle properly).

So, it's not because there's a problem with the status quo that we should ignore it and keep moving on. What I'd like to see before this is ready, probably in some other PR, is some work on making sure the skeleton of these ops get closer to OG codegen and that we handle these promotions/extensions properly during CIRGen - this doesn't mean you need to implement logic for every target or ABI, but the usual way should apply: similar skeleton and asserts for what's not done - anything is better than silently miscompilation. With that in, this PR can then land with checks that use the proper target triple to check their behaviors.